src/functions/BaselineCorrection.C

```diff
@@ -0,0 +1,247 @@
+    mooseError(
+        "In BaselineCorrection ", _name, ": The length of time and acceleration data must be > 0.");
+
+  // check that at lease one least squares fit will be applied
+  if (!isParamValid("accel_fit_order") && !isParamValid("vel_fit_order") &&
+      !isParamValid("disp_fit_order"))
+    mooseError("In BaselineCorrection ",
+               _name,
+               ": No values were input for parameters 'accel_fit_order', 'vel_fit_order', or "
+               "'disp_fit_order'. Please specify an integer value from 0 to 9 for at least one "
+               "of these parameters.");
+
+  // apply baseline correction to raw acceleration time history
+  applyCorrection();
+
+  // try building a linear interpolation object
+  try
+  {
+    _linear_interp = libmesh_make_unique<LinearInterpolation>(_time, _adj_accel);
+  }
+  catch (std::domain_error & e)
+  {
+    mooseError("In BaselineCorrection ", _name, ": ", e.what());
+  }
+}
+
+Real
+BaselineCorrection::value(Real t, const Point & /*P*/) const
+{
+  return _scale_factor * _linear_interp->sample(t);
+}
+
+void
+BaselineCorrection::applyCorrection()
+{
+  // store a reference to final array index
+  unsigned int index_end = _accel.size() - 1;
+
+  // Compute unadjusted velocity and displacment time histories
+  Real dt;
+  std::vector<Real> vel, disp;
+  vel.push_back(0);
+  disp.push_back(0);
+  for (unsigned int i = 0; i < index_end; ++i)
+  {
+    dt = _time[i + 1] - _time[i];
+
+    vel.push_back(BaselineCorrectionUtils::newmarkGammaIntegrate(
+        _accel[i], _accel[i + 1], vel[i], _gamma, dt));
+    disp.push_back(BaselineCorrectionUtils::newmarkBetaIntegrate(
+        _accel[i], _accel[i + 1], vel[i], disp[i], _beta, dt));
+  }
+
+  // initialize polyfits and adjusted time history arrays as the nominal ones
+  DenseVector<Real> coeffs;
+  _adj_accel = _accel;
+  std::vector<Real> p_fit, adj_vel = vel, adj_disp = disp;
+
+  // adjust time histories with acceleration fit if desired
+  if (isParamValid("accel_fit_order"))
+  {
+    _order = getParam<unsigned int>("accel_fit_order");
+    coeffs = BaselineCorrectionUtils::getAccelerationFitCoeffs(
+        _order, _adj_accel, _time, index_end, _gamma);
+
+    for (unsigned int i = 0; i <= index_end; ++i)
+    {
+      p_fit = BaselineCorrectionUtils::computePolynomials(_order, coeffs, _time[i]);
+      _adj_accel[i] -= p_fit[0];
+      adj_vel[i] -= p_fit[1];
+      adj_disp[i] -= p_fit[2];
+    }
+  }
+
+  // adjust with velocity fit
+  if (isParamValid("vel_fit_order"))
+  {
+    _order = getParam<unsigned int>("vel_fit_order");
+    coeffs = BaselineCorrectionUtils::getVelocityFitCoeffs(
+        _order, _adj_accel, adj_vel, _time, index_end, _beta);
+
+    for (unsigned int i = 0; i <= index_end; ++i)
+    {
+      p_fit = BaselineCorrectionUtils::computePolynomials(_order, coeffs, _time[i]);
+      _adj_accel[i] -= p_fit[0];
+      adj_disp[i] -= p_fit[2];
+    }
+  }
+
+  // adjust with displacement fit
```

Comment:
Why are there three options for adjusting the fit? That is, accel, vel, and displacement. I am trying to understand the differences between these.

Reply:
@somu15 this I cannot justify in a GitHub comment, I would need to give you a phone call.

Reply:
@somu15 this approach is explained in a paper: [https://journals.sagepub.com/doi/pdf/10.1177/1687814016675534](https://journals.sagepub.com/doi/pdf/10.1177/1687814016675534)

I included this as a reference on the MooseDocs page.